R/get_mol_kegg.R

```diff
@@ -0,0 +1,113 @@
+utils::globalVariables(".data")
+#' Download compound .mol files from KEGG
+#'
+#' Downloads mol files corresponding to individual compounds or compounds in a
+#' pathway from KEGG.
+#'
+#' @param compound_ids character vector of KEGG compound IDs---5 digits
+#'   prepended with a "C".
+#' @param pathway_ids character vector of KEGG pathway IDs---5 digits prepended
+#'   with "map".
+#' @param dir path to a folder to save mol files in.
+#'
+#' @returns a tibble with the columns `compound_ids`, `pathway_ids` (if used),
+#'   and `mol_paths` (paths to downloaded .mol files)
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' get_mol_kegg(compound_ids = c("C16181", "C06074"), dir = tempdir())
+#' get_mol_kegg(pathway_ids = "map00253", dir = tempdir())
+#' }
+get_mol_kegg <- function(compound_ids, pathway_ids, dir){
+  
+  #TODO: implement redownload=FALSE arg
+  
+  if(missing(dir)) stop("`dir` is required")
+  if ((missing(compound_ids) & missing(pathway_ids)) |
+      !missing(compound_ids) & !missing(pathway_ids)) {
+    stop("One of `compound_id` or `pathway_id` are required")
+  }
+  
+  #if compounds are provided
+  if (!missing(compound_ids)) {
+    if (!all(stringr::str_detect(compound_ids, "^[C][:digit:]{5}$"))) {
+      stop("Some compound_ids are not in the correct KEGG format")
+    }
+    fs::dir_create(dir)
+    out_tbl <-
+      tibble::tibble(compound_id = compound_ids) %>% 
+      dplyr::mutate(mol_path = fs::path(dir, .data$compound_id, ext = "mol"))
+  }
+  # if pathways are provided
+  if (!missing(pathway_ids)) {
+    if (!all(stringr::str_detect(pathway_ids, "^[m][a][p][:digit:]{5}$"))) {
+      stop("Some pathway_ids are not in the correct KEGG format")
+    }
+    fs::dir_create(dir, pathway_ids)
+    compound_ids_list <- lapply(pathway_ids, keggGetCompounds)
+    names(compound_ids_list) <- pathway_ids
+    out_tbl <- 
+      tibble::enframe(compound_ids_list, name = "pathway_id", value = "compound_id") %>% 
+      tidyr::unnest(tidyselect::everything()) %>% 
+      dplyr::mutate(mol_path = fs::path(dir, .data$pathway_id, .data$compound_id, ext = "mol"))
+    
+  }
+  
+  # Download mols
+  .get_mol_kegg <- function(compound_id) {
+    #TODO I think the KEGG API can handle up to 10 requests at once separated by
```

Comment:
I got a timeout, is that related to this API limit? 
```
> ex_pathway <- get_mol_kegg(pathway_ids = "map00253", dir = tempdir())
Error in curl::curl_fetch_memory(url, handle = handle) : 
  Timeout was reached: [rest.kegg.jp] Operation timed out after 10002 milliseconds with 0 out of 0 bytes received
```

Reply:
maybe?  I've gotten occasional timeouts as well, but usually just running the function a second time works.  It would be good to handle this TODO eventually (like, before CRAN submission maybe) but I think it works as-is for now.  If you get time-outs consistently I'll take a closer look.

Reply:
I added a `httr2::req_retry(max_tries = 3)` which should help with this issue